### DISCUSSION TITLE
Somehow make it easier to work with direnv .envrc files

### DISCUSSION THREAD

**Discussion by jackmac92** [twpayne/chezmoi#2178] [Ideas]
I like keeping some various `.envrc` files in my dotfiles, cause I think direnv is great. But direnv does require you to allow `.envrc` files after changes for security reasons

I'd really like if there was someway that chezmoi could run `direnv allow $PATH_TO_A_ENVRC_FILE` after it writes the file to the target

Not sure if direnv support should be a direct feature of chezmoi, is there a way to run a callback after writing a file? E.g. in that callback I could check if the file just written was an `.envrc`?

**Comment by bradenhilton** [comment]
Can you achieve this with a [`run_after_`](https://www.chezmoi.io/reference/target-types/#scripts) script?

**Reply by jackmac92** [reply]
somewhat,
I've been having issues with `run_after` scripts in subdirectories not actually running after (they are actually `run_onchange_after`, but `chezmoi target-path` shows that the `_after` is not part of the target file path, so its not an issue with chezmoi not identifying the `_after` prefix)

Some sort of flag or config, which directs chezmoi to output what files it actually updated would probably handle what I'm after

**Reply by twpayne** [reply]
Running any chezmoi command with the `--debug` flag will give you lots of output. The `after_` does not appear as part of the target path as it an attribute which is stripped from the name.

**Comment by jackmac92** [comment]
yea the `after_` in subdirectories is probably just something I'm doing wrong

I was hoping to somehow run a script to approve direnv changes, when my editor runs `chezmoi apply ~/file-I-just-saved`, it would be nice to somehow have chezmoi also run `direnv allow`

I tried having my editor run apply on the parent of the saved file, but it slows things down a good amount sometimes when I just want to apply one file

I won't be terribly disappointed if this is out of scope for chezmoi, just thought I'd suggest

**Reply by bradenhilton** [reply]
`run_onchange_*` means the script itself will be run only if its own content or filename changes. If you need to run `direnv allow` after every `chezmoi apply`, you should remove `onchange_` and simply use `run_after_`.

I'm not familiar with `direnv` but (assuming you're dealing with static paths) I would start with something like this:

`<dir/of/.envrc>/run_after_00-direnv-allow.sh`:

```sh
#!/usr/bin/env sh
direnv allow <path/to/.envrc>
...
```

If you want to get the paths dynamically, you can make the script scan your home directory for `.envrc` files, then pass these paths to `direnv allow`.

If you need to make the script only run after another file changes, you can use `run_onchange_after_00-direnv-allow.sh.tmpl` and adopt a similar approach as seen in "[Run a script when the contents of another file changes](https://www.chezmoi.io/user-guide/use-scripts-to-perform-actions/#run-a-script-when-the-contents-of-another-file-changes)". That is, if something changes after `chezmoi apply`, and that change can compute a new hash, you can use this trick.

**Comment by nandalopes** [comment]
You can achieve this behavior within direnv config with whitelist

See `man direnv.toml`
